samples/bluetooth/peripheral_hids_keyboard/src/main.c

```diff
@@ -0,0 +1,629 @@
+
+	printk("Pairing cancelled: %s\n", addr);
+}
+
+
+static void auth_done(struct bt_conn *conn)
+{
+	printk("%s()\n", __func__);
+	bt_conn_auth_pairing_confirm(conn);
+}
+
+
+static struct bt_conn_auth_cb conn_auth_callbacks = {
+	.passkey_display = auth_passkey_display,
+	.passkey_entry = NULL,
+	.cancel = auth_cancel,
+	.pairing_confirm = auth_done,
+};
+#else
+static struct bt_conn_auth_cb conn_auth_callbacks;
+#endif
+
+/**@brief Function for key press simulation on selected connection
+ *
+ * This function sends key press and release.
+ *
+ * In case a pattern of 'cofFEe' is the p_key_pattern, with pattern_len as 6 and
+ * pattern_offset as 0, the notifications as observed on the peer side would be
+ * 1>    'c', 'o', 'f', 'F', 'E', 'e'
+ * 2>    -  , 'o', 'f', 'F', 'E', 'e'
+ * 3>    -  ,   -, 'f', 'F', 'E', 'e'
+ * 4>    -  ,   -,   -, 'F', 'E', 'e'
+ * 5>    -  ,   -,   -,   -, 'E', 'e'
+ * 6>    -  ,   -,   -,   -,   -, 'e'
+ * 7>    -  ,   -,   -,   -,   -,  -
+ * Here, '-' refers to release, 'c' refers to the key character being
+ * transmitted.
+ * Therefore 7 notifications will be sent.
+ *
+ * @param string    Key codes to be sent.
+ * @param len       Number of buttons to be sent.
+ * @param conn      Connection handler.
+ * @param boot_mode Information if boot mode protocol is selected.
+ *
+ * @return 0 on success or negative error code.
+ */
+int key_press_conn_send(u8_t *string,
+			u8_t len,
+			struct bt_conn *conn,
+			bool boot_mode)
+{
+	int err = 0;
+	u8_t  offset = 0;
+	u8_t  data[INPUT_REPORT_KEYS_MAX_LEN];
+	u32_t shift_btn;
+
+	while (offset <= len) {
+		memset(data, 0, sizeof(data));
+		memcpy(data + SCAN_CODE_POS + offset, string + offset,
+		       len - offset);
+
+		gpio_pin_read(gpio_devs[SHIFT_BUTTON_PORT], SHIFT_BUTTON,
+			      &shift_btn);
+		if (shift_btn == SHIFT_BUTTON_PRESSED) {
+			data[MODIFIER_KEY_POS] |= SHIFT_KEY_CODE;
+		}
+
+		if (boot_mode) {
+			err = hids_boot_kb_inp_rep_send(&hids_obj, conn,
+							data, sizeof(data),
+							NULL);
+		} else {
+			err = hids_inp_rep_send(&hids_obj, conn,
+						INPUT_REP_KEYS_REF_ID,
+						data, sizeof(data),
+						NULL);
+		}
+		if (err) {
+			break;
+		}
+		offset++;
+	}
+	return err;
+}
+
+
+void key_press_send(u8_t *string, u8_t len)
+{
+	if (len > KEYS_MAX_LEN) {
+		printk("Text is too big\n");
+		return;
+	}
+
+	for (size_t i = 0; i < CONFIG_BT_GATT_HIDS_MAX_CLIENT_COUNT; i++) {
+		if (conn_mode[i].conn) {
+			(void)key_press_conn_send(string, len,
+						  conn_mode[i].conn,
+						  conn_mode[i].in_boot_mode);
+		}
+	}
+}
+
+
+void button_pressed(struct device *gpio_dev, struct gpio_callback *cb,
+		    u32_t pins)
+{
+	static u8_t *chr = hello_world_str;
+
+	if (pins & (1 << SW0_GPIO_PIN)) {
+		key_pack_queue(chr++, 1);
+		if (chr == (hello_world_str + sizeof(hello_world_str))) {
+			chr = hello_world_str;
+		}
+	}
+}
+
+
+void configure_gpio(void)
+{
+	static const u32_t pin_id[4] = { SW0_GPIO_PIN,
```

Comment:
nitpick: break after braces.

Reply:
It is the same in mouse. I prefer to leave it as it would be removed anyway in buttons and leds library. (Where it is placed in the **same fashion** anyway).

Reply:
Ok, not an issue.